src/main/java/baritone/process/elytra/NetherPathfinderContext.java

```diff
@@ -24,98 +24,153 @@
 
     // Visible for access in BlockStateOctreeInterface
     final long context;
     private final long seed;
-    private final ExecutorService executor;
+    // write locked operations
+    private final ExecutorService writeExecutor = Executors.newSingleThreadExecutor();
+    // operations that don't make changes to the chunk cache. could use multiple threads but i'm not sure if it would cause problems.
+    private final ExecutorService readExecutor = Executors.newSingleThreadExecutor();
+    private final ResourceKey<Level> dimension;
 
-    public NetherPathfinderContext(long seed) {
-        this.context = NetherPathfinder.newContext(seed);
+    public NetherPathfinderContext(long seed, Path cache, Level world) {
+        this.dimension = world.dimension();
+        final int dim;
+        if (this.dimension == Level.NETHER) dim = NetherPathfinder.DIMENSION_NETHER;
+        else if (this.dimension == Level.END) dim = NetherPathfinder.DIMENSION_END;
+        else dim = NetherPathfinder.DIMENSION_END;
+        int height = Math.min(world.dimensionType().height(), 384);
+        if (!Baritone.settings().elytraAllowAboveRoof.value && dim == NetherPathfinder.DIMENSION_NETHER) height = Math.min(height, 128);
+        this.context = NetherPathfinder.newContext(seed, cache.toString(), dim, height, Baritone.settings().elytraCustomAllocator.value);
         this.seed = seed;
-        this.executor = Executors.newSingleThreadExecutor();
     }
 
     public boolean hasChunk(ChunkPos pos) {
         return NetherPathfinder.hasChunkFromJava(this.context, pos.x, pos.z);
     }
 
     public void queueCacheCulling(int chunkX, int chunkZ, int maxDistanceBlocks, BlockStateOctreeInterface boi) {
-        this.executor.execute(() -> {
-            synchronized (this.cullingLock) {
+        this.writeExecutor.execute(() -> {
+            writeLock.lock();
+            try {
                 boi.chunkPtr = 0L;
                 NetherPathfinder.cullFarChunks(this.context, chunkX, chunkZ, maxDistanceBlocks);
+            } finally {
+                writeLock.unlock();
             }
         });
     }
 
     public void queueForPacking(final LevelChunk chunkIn) {
         final SoftReference<LevelChunk> ref = new SoftReference<>(chunkIn);
-        this.executor.execute(() -> {
+        this.writeExecutor.execute(() -> {
             // TODO: Prioritize packing recent chunks and/or ones that the path goes through,
             //       and prune the oldest chunks per chunkPackerQueueMaxSize
             final LevelChunk chunk = ref.get();
             if (chunk != null) {
-                long ptr = NetherPathfinder.getOrCreateChunk(this.context, chunk.getPos().x, chunk.getPos().z);
-                writeChunkData(chunk, ptr);
+                writeLock.lock();
+                try {
+                    long ptr = NetherPathfinder.allocateAndInsertChunk(this.context, chunk.getPos().x, chunk.getPos().z);
```

Comment:
I assume this always returns a completely zeroed chunk?

Reply:
Yes